cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocJobMain.java

```diff
@@ -0,0 +1,202 @@
+   * expanded.application.jar/lib/*.jar
+   * expanded.application.jar/classes
+   * expanded.twill.jar
+   * expanded.twill.jar/lib/*.jar
+   * expanded.twill.jar/classes
+   *
+   */
+  private static URL[] getClasspath(URLClassLoader cl, File tempDir) throws IOException {
+    URL[] urls = cl.getURLs();
+    List<URL> urlList = new ArrayList<>();
+    for (String file : ImmutableList.of(Constants.Files.RESOURCES_JAR,
+                                        Constants.Files.APPLICATION_JAR,
+                                        Constants.Files.TWILL_JAR)) {
+      File jar = new File(file);
+      File jarDir = new File(tempDir, "expanded." + file);
+      expand(jar, jarDir);
+      // add url for dir
+      urlList.add(jarDir.toURI().toURL());
+      if (file.equals(Constants.Files.RESOURCES_JAR)) {
+        continue;
+      }
+      urlList.addAll(createClassPathURLs(jarDir));
+    }
+
+    urlList.addAll(Arrays.asList(urls));
+
+    LOG.info("Classpath URLs: {}", Arrays.toString(urlList.toArray()));
+    return urlList.toArray(new URL[0]);
+  }
+
+  /**
+   * Expands jar into destination directory.
+   */
+  private static void expand(File jarFile, File destDir) throws IOException {
+    try (ZipInputStream zipIn = new ZipInputStream(new BufferedInputStream(new FileInputStream(jarFile)))) {
+      Path targetPath = destDir.toPath();
+      Files.createDirectories(targetPath);
+
+      ZipEntry entry;
+      while ((entry = zipIn.getNextEntry()) != null) {
+        Path output = targetPath.resolve(entry.getName());
+
+        if (entry.isDirectory()) {
+          Files.createDirectories(output);
+        } else {
+          Files.createDirectories(output.getParent());
+          Files.copy(zipIn, output);
+        }
+      }
+    }
+  }
+
+  private static List<URL> createClassPathURLs(File dir) {
+    try {
+      List<URL> urls = new ArrayList<>();
+      // add jar urls from lib under dir
+      addJarURLs(new File(dir, "lib"), urls);
+      // add classes under dir
+      urls.add(new File(dir, "classes").toURI().toURL());
+      return urls;
+    } catch (MalformedURLException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static void addJarURLs(File dir, List<URL> result) throws MalformedURLException {
+    Optional<File[]> files = Optional.ofNullable(dir.listFiles(f -> f.getName().endsWith(".jar")));
+    if (files.isPresent()) {
+      for (File file : files.get()) {
+        result.add(file.toURI().toURL());
+      }
+    }
+  }
+
+  /**
+   * Recursively deletes all the contents of the directory and the directory itself.
+   */
+  private static void deleteDirectoryContents(File file) throws IOException {
+    if (file.isDirectory()) {
+      File[] entries = file.listFiles();
+      if (entries != null) {
+        for (File entry : entries) {
+          deleteDirectoryContents(entry);
+        }
+      }
+    }
+    if (!file.delete()) {
+      throw new IOException("Failed to delete " + file);
```

Comment:
For this case, probably better just log instead of throwing. As this could fail the job even the hadoop job succeeded.